### DISCUSSION TITLE
Using Chakra v3 in Turborepo package

### DISCUSSION THREAD

**Discussion by lucasishuman** [chakra-ui/chakra-ui#9301] [Q&A]
I've run into any issue updating my apps to Chakra UI v3 and was wondering if anyone had any ideas.

(note - This same set up **worked** OK with Chakra v2)

Here is my set up:
* A Turborepo (v2.3.3) with a few Next apps (v.14.2.20) and a `@my-app/ui` custom package I created that contains shared components used by all the apps
* I created a Chakra Provider at the top level of each app
* All apps have latest Chakra 3 (v3.2.2) and emotion and next-themes, etc as dependencies

The issue ~
When I import Chakra components directly in an app, or my own components using Chakra which are colocated in the app dir - app loads in browser using `next dev`, no problem.

But when I try to use custom components built with Chakra UI exported from my custom UI package, _or_ the Chakra snippets also located in the package, I see the following error in the browser:
```
ContextError: useContext returned `undefined`. Seems you forgot to wrap component within <ChakraProvider />
```

As if the package components are _not_ running inside the Provider!

I am also able to run a standalone Next.js app using Chakra v3 - where all components and snippets living in same app project - so I know how to config the basics properly.

I know a lot changed in v3 to get Chakra components to work with server or client contexts, so I feel like it may be related to that?

Do I need to export my components from my package in a different way? Or modify the build config of the package somehow?

Thanks for any ideas.

**Comment by lucasishuman** [comment]
If I wrap each component using Chakra in my package with its own ChakraProvider - the package component _will_ appear when imported into an app - but this feels not good for a few reasons...

* each component should not have its own provider context - one top-level provider should be enough (and might cross Chakra wires under the hood? - not entirely sure how context values are used + shared)
* I would have to update all package components to have their own provider
* feels unnecessary and overly complicated (was not necessary in Chakra v2)

**Comment by lucasishuman** [comment]
OK! I think I found a solution.

It seems to have to do with how React Contexts are shared, or not, across apps and packages.

I found that if I move the Charka dependencies in my Turborepo package to be `peerDependencies`, like...

```json
"peerDependencies": {
  "@chakra-ui/react": "^3.2.3",
  "@emotion/react": "^11.14.0",
  "next": "^14.2.20",
  "next-themes": "^0.4.3",
  "react": "^18.3.1",
  "react-dom": "^18.3.1"
}
```

And have all the Chakra dependencies be Turborepo app `dependencies`, like...

```json
"dependencies": {
  "@chakra-ui/react": "^3.2.3",
  "@emotion/react": "^11.14.0",
  "next": "^14.2.20",
  "next-themes": "^0.4.3",
  "react": "^18.3.1",
  "react-dom": "^18.3.1"
}
```

The app can now use the package components that were built using Chakra components and snippets.

I no longer see the `ContextError` in my browser and the UI displays as expected.

**Reply by elisabetta-p** [reply]
Hi @lucasishuman , was that setup enough to not see the error anymore? i have the same issue, with the following versions:

apps/web dependencies
```
  "dependencies": {
    "react": "19.2.0",
    "react-dom": "19.2.0",
    "next": "15.6.0-canary.42",
    "next-themes": "0.4.6",
    "@monorepo/ui": "workspace:*",
    "lodash.isempty": "^4.4.0",
    "@chakra-ui/react": "3.28.0",
    "@emotion/react": "11.14.0"
  },
```

@monorepo/ui package dependencies (chakra-ui is specified only as a peerDependency in this package):
```
"peerDependencies": {
    "@chakra-ui/react": "3.28.0",
    "@emotion/react": "11.14.0",
    "react": "19.2.0",
    "react-dom": "19.2.0",
    "next": "15.6.0-canary.42",
    "next-themes": "0.4.6"
  },
```

The only workaround that i found is this:
- I installed chakra-ui as a dependency in @monorepo/ui
- in @monorepo/ui i import ChakraProvider and then i export it in the index.ts file that exports all the ui custom components
- in the main app, i do `import { ChakraProvider } from '@monorepo/ui'`

Clearly, that workaround is not ideal even if it is working...

**Reply by lucasishuman** [reply]
> was that setup enough to not see the error anymore?

Correct. I can only speak from my specific experience and my project is currently in maintenance mode using React 18, Next 14, Turbo 2.5, + Chakra 3 ~ but what I [described above](https://github.com/chakra-ui/chakra-ui/discussions/9301#discussioncomment-11525476) solved this issue for me. I haven't used Chakra in any other monorepo system since.
